docs/tech-specs/basic-spec.md

```diff
@@ -6,6 +6,256 @@
+    * `X-Broker-Producer-ID` ID of the Producer
+    * `X-Broker-Producer-Token` Token of the Producer
+  * The POST call also will need the channel token for simple auth purpose - `X-Broker-Channel-Token`
+  * The message priority will also need to be passed via header - `X-Broker-Message-Priority`
+  * Message's content type will be derived from request `Content-Type`; if missing will default to `application/octet-stream`
+* The **Message** list endpoint can be filtered using _statusChangedSince_ query parameter
+* The **Message** `GET` endpoint will list all the jobs and their status in the resource itself since the **Message** and **DeliverJob** are both immutable through the API.
+  * **DeliveryJob** will expose re-trigger endpoint to be used by consumer, but only if the delivery status is _dead_.
+* **Message** delivery or **DeliveryJob** can be re-triggered using _Admin Token_ by the Fail-safe worker
+  * That means they will have special re-trigger endpoint which will need _Admin Token_.
+  * DLQ'd jobs can be re-triggered by consumer using its _Consumer Token_.
+  * The Admin token is - `X-Broker-Admin-Token`
+  * The delivery to a consumer will also contain `X-Broker-Consumer-Token` to ensure the request is coming from Broker, in addition to a custom `User-Agent`
+
+So the endpoints available would be -
+
+1. PUT /producer/{producer-id}
+1. GET /producers (Query Params - size, first)
+1. GET /producer/{producer-id}
+1. PUT /channel/{channel-id}
+1. GET /channels (Query Params - size, first)
+1. GET /channel/{channel-id}
+1. PUT /channel/{channel-id}/consumer/{consumer-id}
+1. GET /channel/{channel-id}/consumers (Query Params - size, first)
+1. GET /channel/{channel-id}/consumer/{consumer-id}
+1. POST /channel/{channel-id}/broadcast
+1. GET /channel/{channel-id}/messages (Query Params - statusChangedSince, size, first)
+1. GET /channel/{channel-id}/message/{message-id}
+1. POST /channel/{channel-id}/message/{message-id}/re-trigger (fail-safe only)
+1. POST /channel/{channel-id}/message/{message-id}/job/{job-id}/re-trigger (DLQ or fail-safe)
+
+### Fail-safe worker
+
+Whether the fail-safe worker is working, that will be a configuration; similarly the _Admin Token_ will also be a configuration. Fail-safe worker will simply re-trigger a state using Load-Balanced endpoint of the broker (that is call itself) so that rest of the process works as usual. The spinning up of fail-safe worker will be configurable so that fail-safe is executed from a single execution-unit.
+
+### Libraries
+
+* Schema Migration Management - [GoLang Migrate](https://github.com/golang-migrate/migrate)
+* Web Controller Router - [Gorilla Mux](https://github.com/gorilla/mux) and [Gorilla Handlers](https://github.com/gorilla/handlers)
+* Log Management & Rotation - [Lumberjack.V2](https://github.com/natefinch/lumberjack)
+* Build Time IoC - [Google Wire](https://github.com/google/wire)
+* Configuration parsing - [Go INI](https://github.com/go-ini/ini)
```

Comment:
I would recommend to go with https://github.com/spf13/viper. A very versatile tool for config parsing from JSON, TOML, YAML, HCL, environment variables, command line args, etcd, Consul etc. Will be handy when packaging for cloud native deployments.

Reply:
I feel `viper` does too much, I much prefer INIs simplicity :)